### ISSUE TITLE
@JacksonXmlRootElement malfunction when using it with XmlMapper in multi-thread environment

### ISSUE THREAD

**Issue by wanglingsong** [FasterXML/jackson-dataformat-xml#171]
This is the sample to reproduce the bug. The "localName" override is failed in some of thread, printing original class name "TestModel" instead of "model". XmlMapper seems not thread safe.

```java
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement;

@JacksonXmlRootElement(localName = "model")
public class TestModel {


    public static void main(String[] s) throws Exception {

        final ObjectMapper xmlMapper = new XmlMapper();

        for (int i=0;i < 100;i++) {
            new Thread(new Runnable() {
                @Override
                public void run() {
                    try {
                        System.out.println(xmlMapper.writeValueAsString(new TestModel()));
                    } catch (JsonProcessingException e) {
                        e.printStackTrace();
                    }
                }
            }).start();
        }

        Thread.sleep(10000L);

    }

}
```

**Output:**
```
<model/>
<model/>
<model/>
<model/>
<model/>
<model/>
<model/>
<TestModel/>
<model/>
<model/>
<model/>
<model/>
```

**Comment by cowtowncoder** [comment]
Whoa. That is interesting... I assume this is with 2.6.3?

**Comment by wanglingsong** [comment]
Yes

**Comment by cowtowncoder** [comment]
I can reproduce this. No obvious cause, looking at code, but should not be too difficult to find.

**Comment by cowtowncoder** [comment]
Ok. Yes, interesting. Bug comes from a combination of factors, and was introduced in 2.5, most likely.
Basically, a subset of `AnnotatedClass` instances (ones for which only class annotations are checked, like for use with root name discovery) may be cached (start with 2.5.0). Further, class annotation lookup is actually done lazily. And finally, since (earlier) assumption was that access to `AnnotatedClass` is not multi-threaded, this becomes problematic.

I'll try to get this fixed; 2.6 probably needs bit different fix than what may be done for 2.7, but both should be addressable.
